### ISSUE TITLE
bootstrap: (echo.echo.) was unexpected at this time.

### ISSUE THREAD

**Issue by btm** [chef/knife-windows#211]
@ksubrama can you look at this error and see if it's related to the msi_url work you did in #206? Otherwise we think it's likely winrm 1.3 related:

```
PS C:\Users\btm_000\Documents\GitHub\knife-windows> knife bootstrap windows winrm 192.168.5.109 -x Administrator -P 'P@ssw0rd' -p 80

Waiting for remote response before bootstrap.192.168.5.109 .
192.168.5.109 Response received.
Remote node responded after 0.07 minutes.
Bootstrapping Chef on 192.168.5.109
192.168.5.109 "Rendering "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" chunk 1"
192.168.5.109 "Rendering "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" chunk 2"
192.168.5.109 "quot;^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Set objFSO = CreateObject^^^("Scripting.FileSystemObject"^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.If objFSO.Fileexists^^^(path^^^) Then objFSO.DeleteFile path) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.objFSO.CopyFile sourcePath, path, true) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Set objFSO = Nothing) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Else) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Set objXMLHTTP = CreateObject^^^("MSXML2.ServerXMLHTTP"^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Set wshShell = CreateObject^^^( "WScript.Shell" ^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.Set objUserVariables = wshShell.Environment^^^("USER"^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.rem http proxy is optional) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.rem attempt to read from HTTP_PROXY env var first) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.On Error Resume Next) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.If NOT ^^^(objUserVariables^^^("HTTP_PROXY"^^^) = ""^^^) Then) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.proxy = objUserVariables^^^("HTTP_PROXY"^^^)) && >> "C:\Users\ADMINI~1\AppData\Local\Temp\bootstrap-14096-1426714822.bat" (echo.echo.) was unexpected at this time.
ERROR: Failed to execute command on 192.168.5.109 return code 1
```

**Comment by ksubrama** [comment]
I wasn't able to test the full-pipeline rendering of that vbscript - so it may be due to some escaping related issue.  The error seems to not be that (echo.echo.) was unexpected.  It's actually that "quot;^^^)) && .....  the entire line is unexpected.  This is a part of the script I added.  I tested the raw vbscript but that seemed to work out ok.

Here's the relevant section:
```
sourcePath = Replace(sourcePath, "/", "\\")

Set objFSO = CreateObject("Scripting.FileSystemObject")
```

I suspect that the part after the backslash there is somehow getting translated into:
quot;^^^)).  I suspect that escape_and_echo is not correctly escaping all the characters it needs to escape.

http://ss64.com/nt/syntax-esc.html says that when piping or redirecting (which is happening twice here - once in the template when the vbscript is blitted out and once in the bat that generates the template), \ needs to be escaped.  I believe that this \ causes cmd to reject everything after it.

While drilling into this, I discovered a few other bugs in the windows bootstrap rendering code.  We sometimes render commands longer than 2047 characters long but have simply managed to avoid doing it by pure chance.  I'll take this one and fix both these tomorrow.

**Comment by ksubrama** [comment]
This goes even deeper than I expected.

I deliberately changed the vbs file text to:
```
LALALALALALALALALALALALALALALALALALALALALA
sourcePath = Replace(sourcePath, "/", "\\")
HAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAH
```

I ran knife with -VV and here's the relevant snippet before the failure:
```
DEBUG: Transport: plaintext
DEBUG: cmd.exe /C echo "Rendering "%TEMP%\bootstrap-60921-1426803436.bat" chunk 3" && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.sourcePath = Mid^^^(url, Len^^^("file:///"^^^) + 1^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Else) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.sourcePath = Mid^^^(url, Len^^^("file:"^^^) + 1^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.End If) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.LALALALALALALALALALALALALALALALALALALALALA) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.sourcePath = Replace^^^(sourcePath, "/", "\"^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.HAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAH) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Set objFSO = CreateObject^^^("Scripting.FileSystemObject"^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.If objFSO.Fileexists^^^(path^^^) Then objFSO.DeleteFile path) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.objFSO.CopyFile sourcePath, path, true) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Set objFSO = Nothing) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Else) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Set objXMLHTTP = CreateObject^^^("MSXML2.ServerXMLHTTP"^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Set wshShell = CreateObject^^^( "WScript.Shell" ^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.Set objUserVariables = wshShell.Environment^^^("USER"^^^)) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.rem http proxy is optional) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.rem attempt to read from HTTP_PROXY env var first) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.On Error Resume Next) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.) && >> "%TEMP%\bootstrap-60921-1426803436.bat" (echo.echo.If NOT ^^^(objUserVariables^^^("HTTP_PROXY"^^^) = ""^^^) Then)
```

I enabled analysis logs for winrm in the event manager (http://blogs.msdn.com/b/wmi/archive/2010/03/16/collecting-winrm-traces.aspx) and looked at the incoming events (event ID 772).

```
SOAP [listener receiving index 2 of 5 total chunks (3000 bytes)] Option></w:OptionSet><w:SelectorSet><w:Selector Name="ShellId">0DBCA70F-E426-4C4F-9FC2-7731A693B744</w:Selector></w:SelectorSet></env:Header><env:Body><rsp:CommandLine><rsp:Command>&quot;cmd.exe /C echo &quot;Rendering &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; chunk 3&quot; &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Mid^^^(url, Len^^^(&quot;file:///&quot;^^^) + 1^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Else) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Mid^^^(url, Len^^^(&quot;file:&quot;^^^) + 1^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.End If) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.LALALALALALALALALALALALALALALALALALALALALA) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Replace^^^(sourcePath, &quot;/&quot;, &quot;&quot;cmd.exe /C echo &quot;Rendering &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; chunk 3&quot; &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Mid^^^(url, Len^^^(&quot;file:///&quot;^^^) + 1^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Else) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Mid^^^(url, Len^^^(&quot;file:&quot;^^^) + 1^^^)) &amp;&a
```
```
SOAP [listener receiving index 3 of 5 total chunks (3000 bytes)] mp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.End If) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.LALALALALALALALALALALALALALALALALALALALALA) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.sourcePath = Replace^^^(sourcePath, &quot;/&quot;, &quot;\&quot;^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.HAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAHAH) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Set objFSO = CreateObject^^^(&quot;Scripting.FileSystemObject&quot;^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.If objFSO.Fileexists^^^(path^^^) Then objFSO.DeleteFile path) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.objFSO.CopyFile sourcePath, path, true) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Set objFSO = Nothing) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Else) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Set objXMLHTTP = CreateObject^^^(&quot;MSXML2.ServerXMLHTTP&quot;^^^)) &amp;&amp; &gt;&gt; &quot;%TEMP%\bootstrap-60921-1426803436.bat&quot; (echo.echo.Set wshShell = CreateObject^^^( &quot;WScript.Shell&quot; ^^^)) &amp;&amp; &gt;&gt; &quot;%T
```

**Comment by ksubrama** [comment]
https://github.com/WinRb/WinRM/issues/131

It was my change that caused this but it's a winrm issue.

**Comment by btm** [comment]
It looks like this went out in winrm v1.3.1 so we should be picking it up now.

With the latest bundle install, I'm getting this error, which looks better....:
```
192.168.5.105 powershell.exe -ExecutionPolicy Unrestricted -NoProfile -NonInteractive -File  C:\chef\wget.ps1 "https://www.chef.io/chef/download?p=windows&pv=2008&m=i686&DownloadContext=PowerShell&v=11" "C:\Users\ADMINI~1\AppData\Local\Temp\chef-client-latest.msi"
192.168.5.105 Warning: Failed to download "https://www.chef.io/chef/download?p=windows&pv=2008&m=i686&DownloadContext=PowerShell&v=11" to "C:\Users\ADMINI~1\AppData\Local\Temp\chef-client-latest.msi"
192.168.5.105 Warning: Retrying download with cscript ...
ERROR: WinRM::WinRMHTTPTransportError: Bad HTTP response returned from server (400).
```

If I fix winrm=1.3.0 in the Gemfile.lock I get:
```
Remote node responded after 0.08 minutes.
Bootstrapping Chef on 192.168.5.105
192.168.5.105 The command line is too long.
ERROR: Failed to execute command on 192.168.5.105 return code 1
```

**Comment by btm** [comment]
Closing this as fixed now. If anyone sees and echo.echo issue, be sure you're using winrm v1.3.1+, and if you are, come find me on the internet to talk about it.
